### DISCUSSION TITLE
Inline Rust code in (*.slint) file.

### DISCUSSION THREAD

**Discussion by evan0greenup** [slint-ui/slint#2947] [Ideas]
Is it possible to write Rust callback for slint component as Rust function or closure in `*.slint` file?

Sometimes it would be more convenient to add non-UI-interactive Rust code (file operation, pure-value computing) in `*.slint` file. It would be impressively efficient for nested component.

**Comment by ogoffart** [comment]
One of the design goal is to keep the .slint language simple so it forces a separation between UI and logic, and that it allow design tool to open and operate with the file. 
This gets complicated if one can embed rust code in a .slint file.

The other problem is that the error reporting would be terrible since the error would be pointed at in the generated code. (This would be better within the slint macro, and i considered it in https://github.com/slint-ui/slint/issues/1714 )

Also difficult to access any rust variable from the outside.

So now the "easier" is to create a `global` with some callback implemented in native code.

**Comment by ericguedespinto** [comment]
I appreciate the separation of ui code and logic code but in some cases it makes sense to customise the generated rust widget struct in the same place as slint code.
My use case / example is a calendar widget. To be able to place day buttons inside a month view widget, I need to be able to calculate the weekday of the first day of the month, the total number of days in that month and which years are leap year's. 
It is possible to hook up callbacks and globals to it but it is clumsy and I assume it needs to be done to every instance of the calendar widget. 
One clean solution would be to have a .rs or .slint_rs file in the same directory as the widget .slint file, with the same filename just different extension, where we could have an impl block in pure rust. This file could be included in the proper order by using include_modules!

**Reply by hunger** [reply]
Sounds like a good job for a model to me: Have it return one entry for each cell in the widget and populate the right cells with the data you need.

**Reply by ales-tsurko** [reply]
I went here after searching and I also need a way to bind rust to slint because of the calendar widget :D Such cases make component libraries hard to reuse/distribute.

**Reply by Vadoola** [reply]
I haven't really looked at this code, so I can't speak for how well it works, but while looking around on github for others uses of slint I stumbled across this https://github.com/klaxa/tracking-viewer which has a calendar. Might give some insight on a way to make it work

**Reply by ales-tsurko** [reply]
The only way to make it work is via callbacks and models. In simple apps like you shared it's easy to do, but in some more complex things you starting to think a lot about all those globals and callbacks - everything feels like it would have bugs, because it's hard to track all those globals implementation/assignment collapses and too hacky. It also makes hard to reuse libraries, which relies on Rust bindings. Actually, because of this, while I had very good developer experience when developing/designing the UI itself, when it came to the logic I considered to rewrite the whole app I've been working on for 2-3 months with some other GUI library. Very frustrating.
